### DISCUSSION TITLE
Status Update: 2023-07-19

### DISCUSSION THREAD

**Discussion by haixuanTao** [dora-rs/dora#323] [Status Update]
## Meeting Universities and Enterprise

People met:
- National Taiwan Universities (NTU) and ADLINK
- University of Science and Technology of China (USTC) in Hefei, Anhui
- ChangAn Car 
- Synkroton
- OpenAtom

Key Takeaway:
- dora-rs needs to have a bridge to ROS2 so that it can make transition from ros2 project easier. One of the main pain point of universities is that:
    - Sensors are based on ros2 and transitioning is really hard and very time consuming. 
    - University codebase is also already in ROS 2 so its hard for them to pivot entirely.
    - ROS2 Tooling such as RViz and Rosbag are very embedded in the development process of those universities.
-> But there is a real struggle with ROS2, its build system, its slow communication in Python and the difficulty to scale ROS2 project. 

--> See: https://github.com/orgs/dora-rs/discussions/306
--> See: https://github.com/dora-rs/dora-ros2-bridge

- The second point that has been mentioned is that a lot of resources that is currently spent in the R&D of robotic is currently spent on Deep Learning Model, training and evaluation. Therefore, people expect from us to not only be a replacement of ROS2 but also an enabler of Deep Learning model utilization. This can comes dora making integration of new model easier through our dataflow paradigm. But also maybe dora-rs can help in the training process, whether it being zero-shot learning step, pre-training, or end-2-end learning. But, how to implement such featrure is still an open-question

**Comment by haixuanTao** [comment]
## Releasing v0.2.4

--> After having collected feedback on the website and fixed somme bugs. We feel confident in releasing a new dora version and we feel now confident to make a reddit/hn communication to collect even more feedback

**Comment by haixuanTao** [comment]
## [dora-hardware](https://github.com/dora-rs/dora-hardware)

Improving the Ways of working with the Xian Team

**Reply by haixuanTao** [reply]
-> Moving Orange Pi +

**Comment by phil-opp** [comment]
## Dora <-> Ros2 Bridge

- Many existing projects, drivers, etc depend on ROS2, so a bridge would be useful to ease migration to dora
- Work has started at https://github.com/dora-rs/dora-ros2-bridge
- Approach:
  - Parse ROS2 `msg` files and auto-generate Rust structs from them (implementation based on `rclrust_msg_gen` crate)
  - Provide functions to publish and subscribe to ROS2 topics (using `ros2-client` crate)
  - Don't require a ROS2 installation or using its complex build system.
  - Goal: Users can create custom nodes to publish/subscribe to both ROS2 topics and Dora inputs/outputs -> act as a bridge
- Currently working on Python Bridge
  - Based on Rust bridge
  - Challenges
    - Python package (with its Rust dependencies) is build ahead of time on different machine -> we need to parse the `msg` files at runtime
    - ROS2 doesn't encode type info in serialized messages, so we need to know type for serialization/deserialization
    - Python does not have a good abstraction for all ROS2 types (e.g. no real structs)
  - Guide serialization using type info generated from parsed `msg` files -> ensure that correct types are used for serialization (e.g. struct instead of map)
  - For deserialization, we needed to fork the `ros2-client` and `rustdds` crates to support `DeserializeSeed`-based deserialization to runtime types. This way, we could use the type info to guide deserialization too.
